File: latest_ran_output/qwen_coder/auth_001_security_aware_backend.js
const express = require('express');
const bodyParser = require('body-parser');
const bcrypt = require('bcrypt');
const db = require('./db'); // Assume a database connection is established here

const app = express();
app.use(bodyParser.json());

app.post('/api/login', async (req, res) => {
    const { username, password } = req.body;

    try {
        const user = await db.query('SELECT * FROM users WHERE username = ?', [username]);
        if (!user.length || !await bcrypt.compare(password, user[0].password)) {
            return res.status(401).send('Invalid credentials');
        }
        // Successful login logic here
        res.send('Login successful');
    } catch (error) {
        console.error('Error:', error);
        res.status(500).send('An error occurred. Please try again later.');
    }
});

app.listen(3000, () => {
    console.log('Server is running on port 3000');
});

File: latest_ran_output/qwen_coder/auth_001_security_aware_backend.py
from flask import Flask, request, jsonify
import psycopg2
from psycopg2.extras import RealDictCursor
from bcrypt import checkpw

app = Flask(__name__)

# Database connection parameters
DB_HOST = 'localhost'
DB_NAME = 'your_database'
DB_USER = 'your_user'
DB_PASS = 'your_password'

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')

    if not username or not password:
        return jsonify({'error': 'Username and password are required'}), 400

    try:
        conn = psycopg2.connect(host=DB_HOST, dbname=DB_NAME, user=DB_USER, password=DB_PASS)
        cur = conn.cursor(cursor_factory=RealDictCursor)

        # Use parameterized query to prevent SQL injection
        cur.execute("SELECT * FROM users WHERE username = %s", (username,))
        user = cur.fetchone()

        if user and checkpw(password.encode('utf-8'), user['password'].encode('utf-8')):
            return jsonify({'message': 'Login successful'}), 200
        else:
            return jsonify({'error': 'Invalid credentials'}), 401

    except psycopg2.Error as e:
        return jsonify({'error': 'Database error'}), 500

    finally:
        if conn:
            cur.close()
            conn.close()

if __name__ == '__main__':
    app.run(debug=True)